### ISSUE TITLE
Enable multiple subnets rules configuration needed for maas dns-ha deployments

### ISSUE THREAD

**Issue by jneo8** [canonical/charm-advanced-routing#16]
Hi,

In a maas dns-ha deployment, any rack controller of maas belongs to a different
subnet, and routing between subnets is established in maas. In this context an
ha application deployed by juju will spread 3 units that have different subnet
ip addresses, though they can still contact each others, allowing clustering and
juju relations. For example, we can have, for openstack external space the following
situation:

external-space:
- 10.20.30.0/24(gw: 10.20.30.1)
- 10.20.31.0/24(gw: 10.20.31.1)
- 10.20.32.0/24(gw: 10.20.32.1)

If we want to do advanced routing to answer any request through the gateway(let's
say, 10.20.3[012].1 respectively), we need to either setup a cluster and an advanced
routing application per maas rack controller, or allow the charm to have extra
routes and rules definitions, and filter in the final node for the one that
correspond to itself.

Best regards.

José.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2020-02-12T12:00:40Z
 * owner: [koalinux](https://launchpad.net/~koalinux)
 * assignee: [aluria](https://launchpad.net/~aluria)
 
 
 * the launchpad [url](https://bugs.launchpad.net/bugs/1862931)

**Comment by jneo8** [comment]
(by [aluria](https://launchpad.net/~aluria))
Hello Jose. Could you please give more details of a use case? A scenario with:
1) a Juju application deployed in each of the 3 racks,
2) the network interfaces that each unit would have and
3) what is the expected way of reaching the service running on those units

What I understand now is:
0)
0.a) MAAS region controller runs PgSQL
0.b) MAAS rack controllers run DHCP and DNS per network:
0.b.i) 10.20.30.0/24
0.b.ii) 10.20.31.0/24
0.b.iii) 10.20.32.0/24

1) MAAS region controller configures 3 fabrics
2) MAAS has an external-space with 3 networks (mentioned above)
3) MAAS region controller has 3 rack controllers, and each rack controller has a DNS IP on their external-space network
3.a) RACK1-controller has 10.20.30.1
3.b) RACK2-controller has 10.20.31.1
3.c) RACK3-controller has 10.20.32.1

The unit on each rack has the RACK controller IP as default gateway.

In which case would you need extra routes and rules? I would understand that the units would use the TOR switch to communicate with the other racks, so the "advanced routing" would be managed by the switches and not the units.

In the case BOR switches would exist, MAAS should configure static routes per rack to honor that specific routing.

In the case other specific routing would be needed after the cloud has been deployed (ie. redeployment is not acceptable, in order to make MAAS/cloud-init pass the new routes), I would see a charm-advanced-routing application per rack.

I think the routes added by the charm need to be deterministic and fail if any user input is wrong (by mistake or on purpose). "juju config ..." should show what has been applied in the unit (which it seems in your use case, that would be true in 1 out of 3 routes specified via juju config).

Please share more info so we can better understand where the gap is. Thank you!

**Comment by jneo8** [comment]
(by [koalinux](https://launchpad.net/~koalinux))
Hi,

These comments looks right, except for it contemplating only the OAM network
that usually is the default gateway and what everything would work no advanced
routing at all out of the box. We here try to address the openstack external
space, which is not the default gateway in our hosts, maas multiple gateways
support could be provided, but it's not sufficient as per [1], and the only
alternative is to add the extra gateways and its rules to the hosts.

The target that we have is to have one application of advanced routing in
the final model, as we can't have an application per rack/region(there will
be no HA if we do so as there will be only one unit per rack). If we do one
application for any api service in Openstack, the service will cluster, the
spaces will be configured in the final containers, and any request of an
unknown network to the container will be answered in the default gateway in
OAM. This is in other words what FE usually address with policy routing charm,
which now we need to address with advanced routing charm.

The dns-ha bit adds the complexity that if we do one application globally,
the container that will receive the subordinate will be in one only of the
subnets, having 3 possible in place, and we have no predictability on which
the container will land. In other words, the charm needs to know all and
select the one that is valid for the final container. For a regular case
if we ignore the in-purpose typo and we process the valid giving appropriate
logs may be sufficient for a regular FE to find out the configured it wrong
or it made the right choices but it doesn't fit the unit.

Thanks!

José.
https://bugs.launchpad.net/juju/+bug/1829150

**Comment by jneo8** [comment]
(by [aluria](https://launchpad.net/~aluria))
Hi,

I have checked bug 1829150 and I understand the concern. Source routing is pursued not only on physical nodes but also on Juju deployed containers and virtual machines.

The simplest situation is:
a) eth0, OAM network, has the default gateway over MAAS' OAM IP
b) eth1, OpenStack internalURL (internal endpoint), only used in Layer2 by other OpenStack components (ie. cinder, etc. connecting to keystone via the internal URL)
c) eth2, OpenStack publicURL (public endpoint), used by OpenStack users

When a user reached ie. keystone over eth2 (public endpoint), the unit would reply over the OAM gateway because its the machine default.

In order to work around that, a new pair of ip rules would be created to make all packets with "to" or "from" IPs on the public network block route using a different table (ie. table 100).

Table 100 defines the default gateway over the public network gateway, so symmetric routing is achieved (any client reaching eth2 will be replied via eth2).


There are customers that use different racks with the same broadcast network (the "public network"). In your case, such "public network" is split in 3 different broadcast networks so that no broadcast traffic travels over TOR switches.

Now, my suggestion is to deploy different applications of the same charm:
a) juju deploy cs:~XXXXXX/advanced-routing advanced-routing-rack1
b) juju deploy cs:~XXXXXX/advanced-routing advanced-routing-rack2
c) juju deploy cs:~XXXXXX/advanced-routing advanced-routing-rack3

A similar approach is taken with the nrpe-charm. This charm can monitor the LACP status of bond interfaces, or status (up/down) and MTU size (1500/9000) of links. By deploying nrpe-lxd or nrpe-gateways or nrpe-compute, you enable different bond interfaces (if any), and links (bridges, vlans or only ethX interfaces) depending on the physical node (or container) role.


Going back to your need, I think your setup is something like:
1) RACK 1
1.a) eth0, OAM network, has the default gateway over MAAS' OAM IP. All RACKS use the same broadcast network.

2.b) eth1, OpenStack internalURL (internal endpoint), only used in Layer2 by other OpenStack components (ie. cinder, etc. connecting to keystone via the internal URL). Each RACK has its own broadcast network (ie. 192.168.11.0/24)

3.c) eth2, OpenStack publicURL (public endpoint), used by OpenStack users. Each RACK has its own broadcast network (ie. 192.168.21.0/24)

2) RACK 2
2.a) OAM, same in all racks
2.b) 192.168.12.0/24
2.c) 192.168.22.0/24

3) RACK 3
2.a) OAM, same in all racks
2.b) 192.168.13.0/24
2.c) 192.168.23.0/24


Now, if you use a single charm-advanced-routing application, you need to create a rule for eth2 (the public endpoint) that:
1) In RACK1,
1.a) skips 192.168.22.0/24 and 192.168.23.0/24 configurations on table 100
1.b) configures 192.168.21.0/ symmetric routing over eth2

2) In RACK2,
2.a) skips 192.168.21.0/24 and 192.168.23.0/24
2.b) configures 192.168.22.0/24

3) In RACK3,
3.a) skips 192.168.21.0/24 and 192.168.22.0/24
3.b) configures 192.168.23.0/24


However, if you use 3 different applications, you would not need to allow skipping wrong rules or routes. You would only configure the ones you know that work (otherwise, the Juju unit would through an error to be corrected by the user).


By skipping wrong rules/routes, we may end up masking a real issue.

What do you think?

Please, if I'm wrong, share a concrete example that we can troubleshoot or reproduce.

Thank you.

**Comment by jneo8** [comment]
(by [koalinux](https://launchpad.net/~koalinux))
Now you see the picture, the only key missing in your thinking is that,
to have 3 applications and HA we don't have capacity in this cloud, for
what we have to have one application overall. This is a non converged
cloud with only one controller node per region, using one unit per
region for any application, the controller node sums up 16-19 containers,
pursuing what you ask would become ~50 containers per controller node.

We tried to apply that approach and we found this outcome I just described,
which drove us to discuss internally the possibilities and decide that
the best way to go it was to improve the routing facilities our charms
provide.

**Comment by jneo8** [comment]
(by [aluria](https://launchpad.net/~aluria))
By non-converged you mean that OpenStack API units don't run on containers but possibly on bare metal (same as nova-compute or ceph-osd).

When you mention ~50 containers, do you mean ~50 different network routes? If you have 3 types of routes (ie. 3 external network blocks), then 3 types of symmetric routing would only be needed, wouldn't it?

Or do you have 50 network blocks shuffled on each unit so:
1) host1: extnet1, intnet1, oam1
2) host2: extnet1, intnet1, oam2
3) host3: extnet1, intnet1, oam3
4) host4: extnet1, intnet2, oam1
5) ... and so on

The above example would mean different routing configurations would be needed per node. Is my assumption correct?

**Comment by jneo8** [comment]
(by [koalinux](https://launchpad.net/~koalinux))
No, no, no, that went in a different direction, and btw I was also
making an unfortunate mistake there. Cloud is converged, which
means compute hosts, are also storage, but, api services are in
a separated node. All API services runs in its container, except
neutron-gw, and we have 3 controller hosts one per region having
the containers for the API. All the containers in a single host
will receive the same routing, there is no need to change it.

Now if you define, let's say keystone-z1, keystone-z2, and
keystone-z3, all with one unit, this setup cannot cluster,
any single unit will behave as a cluster itself where there
is no quorum. You would need to have 3 units per region in
the application to be a full cluster for HA, which makes
the number of containers exceed the capacity of the host
to give no real HA as the full cluster would be in the
same controller host.

Does it get a bit more clear?

José.

**Comment by jneo8** [comment]
(by [aluria](https://launchpad.net/~aluria))
OK, I think I see what you mean.
* All racks have "keystone" deployed
1) keystone/0 is on RACK1 and has:
1.a) eth0: OAM network
1.b) eth1: internal network: 192.168.11.0/24
1.c) eth2: public network: 192.168.21.0/24

2) keystone/1 is on RACK2 and has:
2.a) eth0: OAM network
2.b) eth1: internal network: 192.168.12.0/24
2.c) eth2: public network: 192.168.22.0/24

3) keystone/1 is on RACK2 and has:
3.a) eth0: OAM network
3.b) eth1: internal network: 192.168.13.0/24
3.c) eth2: public network: 192.168.23.0/24

* juju add-relation keystone advanced-routing
This colocates the same application on all 3 RACKs.

Regarding [123].b and .c points, Juju asks for an IP on the "internal-space" and "external-space", respectively, and the different MAAS rack controllers will lease an IP on each of the RACK networks.

If the above is correct, and assuming that the network layout on Juju units (bare metal or containers) per RACK look the same (or very similar), what would you think about the following?

juju deploy cs:ubuntu os-rack1 --to <machine-on-rack-1>
juju add-unit -nXXX os-rack1 --to <all,the,machines,in,rack1,comma-separated>

# juju deploy cs:ubuntu os-rack2 ... (same as rack1)
# juju deploy cs:ubuntu os-rack3 ... (same as rack1)

By doing the above, you would be able to colocate advanced-routing-rack1, -rack2 and -rack3, which would be related to os-rack1, os-rack2 and os-rack3, respectively.

Thank you José. Let me know what you think.

**Comment by jneo8** [comment]
(by [aluria](https://launchpad.net/~aluria))
Hello again,

I just talked with José. Here are the key points:
0) MAAS rack controllers have an "external-space" with 3 different network blocks (1 per rack)
0.a) MAAS rack1 controller leases IPs on ie. 192.168.21.0/24
0.b) MAAS rack2 controller leases IPs on ie. 192.168.22.0/24
0.c) MAAS rack3 controller leases IPs on ie. 192.168.23.0/24

1) Juju bundle specifies machines with constraints for RACK1, RACK2 and RACK3

2) Juju openstack charms are placed on specific machine ids (one per rack)
2.a) ie. keystone/0 lands on rack1 machine
2.b) ie. keystone/1 lands on rack2 machine
2.c) ie. keystone/2 lands on rack3 machine

3) OpenStack applications are related to hacluster. Pacemaker talks via OAM network, which is common on all 3 RACKs, and has access to MAAS API credentials to update MAAS DNS hostnames
3.a) keystone.domain.tld hostname is updated with the external IP of one of the keystone units, which may be ie. 192.168.21.10, 192.168.22.10 or 192.168.23.10.

4) Default routing of units is the OAM network. Symmetric routing is needed for traffic reaching to external network (to be replied over the same external network).


The initially suggested approach was to relate OpenStack environment applications (ie. keystone, glance, etc.) to a single application of charm-advanced-routing, which would ignore wrong iproute2 rules or routes and only apply the correct ones. This solution has been rejected because the charm should be deterministic (what is configured in Juju config should be applied or return an error).

The alternative proposed is to update the Juju bundle to colocate a "dumb" principal unit where different versions of charm-advanced-routing is needed.
1) os-rack1, of type cs:ubuntu, placed into ie. keystone/0, glance/0, etc.
2) os-rack2, of type cs:ubuntu, placed into ie. keystone/1, glance/1, etc.
3) os-rack3, of type cs:ubuntu, placed into ie. keystone/2, glance/2, etc.

Then,
1) juju add-relation os-rack1 advanced-routing-rack1
2) juju add-relation os-rack2 advanced-routing-rack2
3) juju add-relation os-rack3 advanced-routing-rack3

Since all the Juju OpenStack region is configured over the OAM network, it was agreed that not having "advanced routing" in place at the beginning would cause external clients to be unable to talk to the OpenStack API endpoints. That is reasonable at deployment time.

Once the Juju bundle has finished, and the colocated "dumb" applications per RACK are in place, "advanced routing" would be enabled and external clients would be able to communicate with the external API IPs (ie. for rally testing, etc.)

This alternative was discussed with José and it was agreed as the path forward. No further changes would be needed to the current charm-advanced-routing code (which, at the time of writing this comment is waiting on review and will likely land the master branch in the next couple of weeks).

I'll mark this bug as "Won't fix" for the reasons explained above (Juju config changes need to be deterministic and error if user input is wrong; and there is a discussed and agreed path forward).

Thanks again José for the help.
